File: 1391-check-if-there-is-a-valid-path-in-a-grid/1391-check-if-there-is-a-valid-path-in-a-grid.java
class Solution {

    Map<Integer, int[][]> mpp = new HashMap<>();
    int n, m;
    int[][] visited;

    boolean isValid(int i, int j){
        return i >= 0 && i < n && j >= 0 && j < m;
    }

    boolean solution(int i, int j, int[][] grid){
        if(i == n-1 && j == m-1) return true;

        visited[i][j] = 1;

        int[][] directions = mpp.get(grid[i][j]);

        for(int[] dir : directions){
            int dx = i + dir[0];
            int dy = j + dir[1];

            if(!isValid(dx, dy) || visited[dx][dy] == 1) continue;

            // check reverse connection
            boolean ok = false;
            for(int[] back : mpp.get(grid[dx][dy])){
                if(dx + back[0] == i && dy + back[1] == j){
                    ok = true;
                    break;
                }
            }

            if(ok && solution(dx, dy, grid)){
                return true;
            }
        }
        return false;
    }

    public boolean hasValidPath(int[][] grid) {
        mpp.put(1, new int[][]{{0,-1},{0,1}});
        mpp.put(2, new int[][]{{1,0},{-1,0}});
        mpp.put(3, new int[][]{{0,-1},{1,0}});
        mpp.put(4, new int[][]{{0,1},{1,0}});
        mpp.put(5, new int[][]{{0,-1},{-1,0}});
        mpp.put(6, new int[][]{{0,1},{-1,0}});

        n = grid.length;
        m = grid[0].length;

        visited = new int[n][m];

        return solution(0, 0, grid);
    }
}

File: 1391-check-if-there-is-a-valid-path-in-a-grid/1391-check-if-there-is-a-valid-path-in-a-grid.cpp
class Solution {
public:

    unordered_map<int,vector<vector<int>>>mpp;
    int n,m;
    vector<vector<int>>visited;

    bool isValid(int i,int j){
        return (i>=0 && i<n) && (j>=0 && j<m);
    }

    bool solution(int i,int j,vector<vector<int>>&grid){
        if(i==n-1 && j==m-1) return true;
        if(i>=n && j>=m) return false;
        visited[i][j]=1;

        auto possibleDirections=mpp[grid[i][j]];

        for(auto direction:possibleDirections){
            int dx=i+direction[0];
            int dy=j+direction[1];

            if(!isValid(dx,dy) || visited[dx][dy]) continue;

            bool ok = false;
            for(auto back : mpp[grid[dx][dy]]){
                if(dx + back[0] == i && dy + back[1] == j){
                    ok = true;
                    break;
                }
            }

            if(ok && solution(dx,dy,grid)){
                return true;
            }
        }
        return false;
    }

    bool hasValidPath(vector<vector<int>>& grid) {
        mpp[1]={{0,-1},{0,1}};
        mpp[2]={{1,0},{-1,0}};
        mpp[3]={{0,-1},{1,0}};
        mpp[4]={{0,1},{1,0}};
        mpp[5]={{0,-1},{-1,0}};
        mpp[6]={{0,1},{-1,0}};

        n=grid.size();
        m=grid[0].size();
        visited.assign(n,vector<int>(m,0));

        return solution(0,0,grid);
    }
};